polynote-frontend/polynote/ui/component/notebook/cell.ts

```diff
@@ -1551,3 +1562,291 @@ export class TextCell extends Cell {
         this.editor.disabled = disabled
     }
 }
+
+
+export class VizCell extends Cell {
+
+    private editor: VizSelector;
+    private editorEl: TagElement<'div'>;
+    private cellInputTools: TagElement<'div'>;
+    private execDurationUpdater: number;
+    private viz: Viz;
+    private cellOutput: CodeCellOutput;
+
+    private valueName: string;
+    private resultValue?: ResultValue;
+    private previousViews: Record<string, [Viz, Output | ClientResult]> = {};
+
+    constructor(dispatcher: NotebookMessageDispatcher, notebookState: NotebookStateHandler, cellState: StateHandler<CellState>, private path: string) {
+        super(dispatcher, notebookState, cellState);
+
+        const initialViz = parseMaybeViz(this.cellState.state.content);
+        if (isViz(initialViz)) {
+            this.viz = initialViz;
+            this.valueName = this.viz.value;
+        } else if (initialViz && initialViz.value) {
+            this.valueName = initialViz.value;
+        } else {
+            throw new window.Error("No value defined for viz cell");
+        }
+
+
+        this.editorEl = div(['viz-selector', 'loading'], 'Notebook is loading...');
+
+        if (cellState.state.output.length > 0) {
+            this.previousViews[this.viz.type] = [this.viz, cellState.state.output[0]];
+        }
+
+        // TODO: extract the common stuff with CodeCell
+
+        const execInfoEl = div(["exec-info"], []);
+
+        this.cellOutput = new CodeCellOutput(dispatcher, cellState, StateHandler.from(undefined, this), StateHandler.from(undefined, this));
```

Comment:
No possibility of errors?

Reply:
There could be runtime errors. I'll add handling for that in a subsequent PR (don't want to change too much in this PR, because I have more PRs I have to rebase onto this one which were based on #994)

Reply:
👍  SGTM! can you add a TODO so we don't forget? 😉